File: R/make_report.R
# Generate a standalone HTML season report (R instance).
#
#   Rscript R/make_report.R [league_id] [--season 2024] [--out file.html] [--all]
#
# Invoked by launch.py ("r report"); loads the package via pkgload and writes the
# report(s) to the repo root. Mirrors python/make_report.py.

suppressWarnings(suppressMessages(pkgload::load_all("R/sleepermetrics", quiet = TRUE)))

args <- commandArgs(trailingOnly = TRUE)
get_opt <- function(flag) {
  i <- match(flag, args)
  if (is.na(i) || i == length(args)) NULL else args[[i + 1]]
}
positional <- args[!grepl("^--", args) &
                     !(seq_along(args) %in% (match(c("--season", "--out"), args) + 1))]
league <- if (length(positional)) positional[[1]] else
  Sys.getenv("SLEEPERMETRICS_LEAGUE", "1252770181306929152")
season_arg <- get_opt("--season")
out_arg <- get_opt("--out")
all_seasons <- "--all" %in% args
playoff_dir <- Sys.getenv("SLEEPERMETRICS_PLAYOFFS", "playoffs")

ss <- sl_apply_playoffs(sl_seasons(league), playoff_dir)
if (!length(ss)) stop("No scored seasons found for league ", league)
ids <- vapply(ss, function(s) s$league_id, character(1))
pos <- sl_load_playoffs(playoff_dir, league_ids = ids)

targets <- if (all_seasons) names(ss) else {
  if (!is.null(season_arg) && season_arg %in% names(ss)) season_arg else names(ss)[length(ss)]
}
for (key in targets) {
  out <- if (!is.null(out_arg) && !all_seasons) out_arg else
    sprintf("report_%s_%s.html", league, key)
  sl_season_report(ss[[key]], out, seasons = ss, playoffs = pos)
  cat(sprintf("wrote %s  (%s %s)\n", out, ss[[key]]$name, key))
}

File: python/make_report.py
"""Generate a standalone HTML season report (Python instance).

    python launch.py report                     # latest season, default league
    python launch.py report <league_id>
    python launch.py report <league_id> --season 2024
    python launch.py report <league_id> --out my_report.html
    python launch.py report <league_id> --all   # one file per season

Run via launch.py (which sets the venv + season dir), or directly from python/.
"""
from __future__ import annotations

import argparse
import os
import sys

import sleepermetrics as sm

DEFAULT_LEAGUE = os.environ.get("SLEEPERMETRICS_LEAGUE", "1252770181306929152")
SEASON_DIR = os.environ.get("SLEEPERMETRICS_SEASON_DIR", "season")


def main(argv=None) -> int:
    ap = argparse.ArgumentParser(prog="make_report")
    ap.add_argument("league", nargs="?", default=DEFAULT_LEAGUE)
    ap.add_argument("--season", help="season to report (default: latest)")
    ap.add_argument("--out", help="output path (default: report_<league>_<season>.html)")
    ap.add_argument("--all", action="store_true", help="one report per season")
    a = ap.parse_args(argv)

    ss = sm.apply_playoffs(sm.seasons(a.league), SEASON_DIR)
    pos = sm.load_playoffs(SEASON_DIR, league_ids=[s.league_id for s in ss.values()])
    if not ss:
        sys.exit(f"No scored seasons found for league {a.league}.")

    targets = list(ss) if a.all else [a.season if a.season in ss else list(ss)[-1]]
    for key in targets:
        out = a.out if (a.out and not a.all) else f"report_{a.league}_{key}.html"
        sm.season_report(ss[key], out, seasons=ss, playoffs=pos)
        print(f"wrote {out}  ({ss[key].name} {key})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
